### PR TITLE
[Clover step-11] Set the validation

### PR DESCRIPTION
**Pull request by clover-zhao**
### Summary/概要
Description:

This PR enhances the overall robustness of the application and improves the user experience by ensuring that only valid data can be entered for tasks.

Changes Made:

In this PR, I added validations to the Task model to enforce data integrity and improve data quality.
Furthermore, I updated the view templates to display informative validation error messages when users attempt to create or update a task with invalid data. 
Additionally, I implemented a comprehensive model test suite to thoroughly test the validations. 

### How to check the behavior/動作確認
- Display error message
  
  Screenshot 2023-08-03 at 10 23 57

- Test result
  image

### CODE REVIEW

File: myapp/app/models/task_label.rb [hunk 1]
```diff
@@ -1,4 +1,7 @@
 class TaskLabel < ApplicationRecord
+  validates :task_id, presence: true
+  validates :label_id, presence: true
```
Comment (anneju) [inline myapp/app/models/task_label.rb]: you don't need these as you already has the `belong_to` association

File: myapp/app/views/tasks/_form.html.erb [hunk 2]
```diff
@@ -34,11 +34,12 @@
...
-
-  <div class="field">
+  
+  <%# todo: add labels to tasks %>
+  <%# <div class="field">
     <%= form.label :label_ids, t('task.table_headers.labels') %>
-    <%= form.collection_select :label_ids, @labels, :id, :name, {}, multiple: true %>
-  </div>
+    <%# <%= form.collection_select :label_ids, @labels, :id, :name, {}, multiple: true %> <%# %>
+  <%# </div> %> <%# %>
```
Comment (anneju) [inline myapp/app/views/tasks/_form.html.erb]: sorry, what was this that you commented out? @@

**Review by anneju — COMMENTED** [review]
The point of this task https://github.com/Fablic/training/blob/develop/steps_en.md#step-11-lets-set-the-validation is to

> create a migration that also sets DB constraints

Rails validation cannot be trusted, only DB level constraints is guaranteed.

File: myapp/app/models/task_label.rb (hunk 1 above)
Comment (clover-zhao) [reply myapp/app/models/task_label.rb]: Thanks for this information, I will remove this two lines

**Comment by clover-zhao** [comment]
> The point of this task https://github.com/Fablic/training/blob/develop/steps_en.md#step-11-lets-set-the-validation is to
> 
> > create a migration that also sets DB constraints
> 
> Rails validation cannot be trusted, only DB level constraints is guaranteed.

I am not sure if I did it in a right way, I have set some constraints when I create Task table, please refer this  [`db/migrate/20230719025223_create_tasks.rb`](https://github.com/Fablic/training/blob/clover-step11/myapp/db/migrate/20230719025223_create_tasks.rb) file.
```
class CreateTasks < ActiveRecord::Migration[6.0]
  def change
    create_table :tasks do |t|
      t.references :user, null: false, foreign_key: true
      t.string :name, null: false, default: 'New Task'
      t.text :description
      t.integer :priority, null: false, default: 0
      t.date :deadline
      t.string :status, null: false, default: 'Not Started'
      t.timestamps
    end
  end
end
```

File: myapp/app/views/tasks/_form.html.erb (hunk 2 above)
Comment (clover-zhao) [reply myapp/app/views/tasks/_form.html.erb]: This part is for labels function, I will add it in the future. It is better to remove those in current step?

Comment (anneju) [reply myapp/app/views/tasks/_form.html.erb]: it's fine, just for this kind of things, you can add a comment nearby so we know what it is and why that you want to comment it out for now.

**Comment by anneju** [comment]
@clover-zhao Thanks! I didn't noticed you have the migration files created before.

**Review by anneju — COMMENTED** [review]
could you also add model specs for other models? like `user`, `label`

File: myapp/app/views/tasks/_form.html.erb (hunk 2 above)
Comment (clover-zhao) [reply myapp/app/views/tasks/_form.html.erb]: Sure, I have a todo comment.

**Comment by clover-zhao** [comment]
> could you also add model specs for other models? like `user`, `label`

Sure, I have add them.
Screenshot 2023-08-04 at 10 44 05

File: myapp/spec/factories/label.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+FactoryBot.define do
+  factory :label do
+    name { 'test' }
+  end
+end
```
Comment (anneju) [inline myapp/spec/factories/label.rb]: is that an extra blank line at bottom?

File: myapp/spec/models/label_spec.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe Label, type: :model do
+  describe "validations" do
+    it "is valid with valid attributes" do
```
Comment (anneju) [inline myapp/spec/models/label_spec.rb]: it's better to wrap with a `context` 

```
describe "validations" do
  context 'when valid' do 
    ....
  end

  context 'when not valid' do 
    ....
  end
end
```

**Review by anneju — COMMENTED** [review]
some comments, also a good material of spec https://www.betterspecs.org/ for your reference~

File: myapp/spec/factories/label.rb (hunk 3 above)
Comment (clover-zhao) [reply myapp/spec/factories/label.rb]: Fixed ^o^

File: myapp/spec/models/label_spec.rb (hunk 4 above)
Comment (clover-zhao) [reply myapp/spec/models/label_spec.rb]: Sure, I have wrap them with a `context`

**Comment by clover-zhao** [comment]
> some comments, also a good material of spec https://www.betterspecs.org/ for your reference~

THX, for this information, I will check this website for learning more about how to write a good spec.

File: myapp/spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe "associations" do
```
Comment (anneju) [inline myapp/spec/models/user_spec.rb]: why `associations` here? typo?

File: myapp/spec/models/task_spec.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
+RSpec.describe Task, type: :model do
+  before do
+    @user = create(:user)
+  end
```
Comment (anneju) [inline myapp/spec/models/task_spec.rb]: what is this `@user` for?

**Review by anneju — COMMENTED** [review]
some questions~

File: myapp/spec/models/task_spec.rb (hunk 6 above)
Comment (clover-zhao) [reply myapp/spec/models/task_spec.rb]: Task is associate with user, so need create a user before. But no need for define a user variable. I have remove it.

File: myapp/spec/models/user_spec.rb (hunk 5 above)
Comment (clover-zhao) [reply myapp/spec/models/user_spec.rb]: Yeah, thanks for notice this.

File: myapp/spec/models/task_spec.rb (hunk 6 above)
Comment (anneju) [reply myapp/spec/models/task_spec.rb]: in factory bot, you can create a associated record with `before`, `after`... ref: https://stackoverflow.com/a/71482271

Please try to use that and remove unneeded user creation.

**Review by anneju — COMMENTED** [review]
maybe a little enhancement..?

File: myapp/spec/models/task_spec.rb (hunk 6 above)
Comment (clover-zhao) [reply myapp/spec/models/task_spec.rb]: I have refactor the `user` and  `task` model spec, please have a look.

**Review by anneju — APPROVED** [review]
looks good~

**Pull request merged**